### DISCUSSION TITLE
Could you please explain why the default hashIterations is set to 21000?

### DISCUSSION THREAD

**Discussion by ghost** [keycloak/keycloak#29590] [Q&A]
Could you please provide an explanation? I've noticed that during load testing, the CPU is experiencing significant strain.
![1](https://github.com/keycloak/keycloak/assets/163235135/ecbe7203-7271-4344-8e1e-0413d92846e2)

**Comment by justin-tay** [comment]
Those values are from the [OWASP Password Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2)

See also
- https://tobtu.com/minimum-password-settings/
- https://www.nist.gov/news-events/news/2023/05/nist-revise-sp-800-132-recommendation-password-based-key-derivation-part-1
- https://csrc.nist.gov/csrc/media/Projects/crypto-publication-review-project/documents/initial-comments/sp800-132-initial-public-comments-2023.pdf
